New package: FMM3D v1.0.1

JuliaRegistrator opened the pull request:
- Registering package: FMM3D
- Repository: https://github.com/flatironinstitute/FMM3D
- Created by: @lu1and10
- Version: v1.0.1
- Commit: b0dc85a3da10257adedd32b6068c8b242a6cecbc
- Reviewed by: @lu1and10
- Reference: https://github.com/flatironinstitute/FMM3D/commit/b0dc85a3da10257adedd32b6068c8b242a6cecbc#commitcomment-139170329
- Description: Flatiron Institute Fast Multipole Libraries --- This codebase is a set of libraries to compute N-body interactions governed by the Laplace and Helmholtz equations, to a specified precision, in three dimensions, on a multi-core shared-memory machine. 
<!-- daf33ad5e6d085e55119bc06703763cec7c0b68dd6e421de144489cbdea59452f02bd83fe8a184914156eaf513823d4da5d031214822f9e9b2d69203503e54631078c3f23ce2146c5bded6b04ca9492b0014703b2e5ad41bf44f1cc130e3fd8bcf0624ab4aca34d2c8add5dbc714992f23ea9c46efe78fbbf165760f52e4cb85f5fbe4371d1850af563f5a7530dd4bad8752983b01a131ab92bc29b552b47779140a855561d0540c45bdfcd9e5f818a0e8b0d9337014daa8cc3314bd2e458c7c67eeb69e09fc33865bfb3cae1b861bf514ac03629b098a2d224e177c737653e2 -->

mipals commented:
This package is a wrapper and is therefore named after the library that it wraps.

The [original library has an Apache 2.0 license](https://github.com/flatironinstitute/FMM3D/blob/master/LICENSE). This is registration is only of the subfolder including the Julia wrapper which does not contain the license. Is this the reason why it complains regarding the license?

Regarding the name, then FMM2D is for 2D computations while FMM3D for the 3D computations. I think the differences is quite clear.

shayandavoodii commented:
[noblock]

Regarding the name, what if you combine the FMM2D and the FMM3D in a package called FMMS or whatever? I mean, instead of having these two packages, let's have a package that hopefully includes all the variants.

mipals commented:
I do not really know what is the best practice here. The two packages are wrappers to two different (but related) libraries, which would be indicated by having two separate packages. That being said I am not against having a combined package.

lu1and10 commented:
The 2D and 3D library function signatures seems to be quite similar. If @askhamwhat and @mipals both think merging the 2D and 3D is a good idea, I guess we should start a new repo and collaborate on combining the two wrappers. @askhamwhat what do you think?

mipals commented:
The 2D wrapper is indeed very similar (in fact it is very much just a copy-paste of the 3D wrapper with a few things changed). Note that the 2D wrapper was written quite quickly and could use a freshener at some point. In fact there are many features that it does on include as only the Helmholtz kernel was covered in the documentation at the time it was written (I do know if this has since been fixed). 

Whether or not it makes sense to have a single package I do not know. Part of my likes that they are separated (due to the original packages being separate), but I can also see the idea combining them working. 

Regarding the license issue then the reason is that only the subfolder is distributed. Frederik Ekre therefore suggested on Slack to copy the license to the Julia subfolder. I guess this should be possible?

lu1and10 commented:
> Regarding the license issue then the reason is that only the subfolder is distributed. Fredrik Ekre therefore suggested on Slack to copy the license to the Julia subfolder. I guess this should be possible?

Yes, @mrachh is working on editing and update the license for FMM3D, once it's done we will make a copy to the subfolder.

maltezfaria commented:
To weight in my opinion as a user of both libraries: 
- Despite the one character difference, the distinction between `FMM3D` vs `FMM2D` is quite clear
- Since `FMM2D` and `FMM3D` wrap different libraries, I think they should exist as different Julia packages. 

Thanks for trying to register this!

tanderson92 commented:
> To weight in my opinion as a user of both libraries:
> 
> * Despite the one character difference, the distinction between `FMM3D` vs `FMM2D` is quite clear
> * Since `FMM2D` and `FMM3D` wrap different libraries, I think they should exist as different Julia packages.
> 
> Thanks for trying to register this!

I am also a user of both libraries (and recently wrapped each for a project, so I'm familiar with the differences) and I would second this perspective. The difference is clear.

Moreover, wrapping them in the same way risks confusion at important implementational differences: for example, while the interface may be similar, the 3D and 2D libraries differ significantly in how they treat evaluation at source-points (targets having a nontrivial intersection with sources). Thus, in the end the user is going to need to pay heed to the documentation for the relevant library, anyway.

mipals commented:
Yeah, I am also thinking that having two separated packages is probably the way to go due to them wrapping two different libraries. No reason to mix them up.  

@tanderson92 A little off topic but are you sure that the FMM2D does not support Laplace? I thought it was just undocumented (similar to Stokes and Biharmonic) and hence did not include it in the Julia wrapper.

tanderson92 commented:
> Yeah, I am also thinking that having two separated packages is probably the way to go due to them wrapping two different libraries. No reason to mix them up.
> 
> @tanderson92 A little off topic but are you sure that the FMM2D does not support Laplace? I thought it was just undocumented (similar to Stokes and Biharmonic) and hence did not include it in the Julia wrapper.

I was speaking only to the Julia wrapper, which is still relevant for this discussion. Yes, the underlying library has more capabilities that aren't currently exposed in `FMM2D.jl`.

mipals commented:
Fair enough @tanderson92. If you at some point need the Julia wrapper of FMM2D to also include Laplace, Stokes, etc. let me know. I just stopped at Helmholtz since that was what I needed (and it was the only thing documented) and since then nobody else than me has seemed to be using the wrapper 😄

mipals commented:
@DilumAluthge There is a now a license in the Julia subfolder. Can the automerge be run again?

DilumAluthge commented:
> Can the automerge be run again?

Yep, just re-trigger Registrator on your latest commit.

[noblock]

JuliaRegistrator committed:
New package: FMM3D v1.0.1

UUID: 1e13804c-f9b7-11ea-0ef0-29f3b1745df8
Repo: https://github.com/flatironinstitute/FMM3D.git
Tree: d53f013f8b1830537322729258a4dadb1d4c3a51

Registrator tree SHA: 17aec322677d9b81cdd6b9b9236b09a3f1374c6a

DilumAluthge commented:
> There is a now a license in the Julia subfolder. Can the automerge be run again?

[noblock]

Looks like the license is fixed now!

maltezfaria commented:
Auto-merge fails due to the *similar name* conflict (`FMM2D` and `FMM3D`). 

@DilumAluthge can this be manually merged now? Thanks!

DilumAluthge commented:
Would you be open to a longer and more descriptive name?

mipals commented:
I agree completely with you @maltezfaria  👍

DilumAluthge commented:
@lu1and10 As the package author, would you be open to a longer and more descriptive name?

We usually try for package names that are descriptive to the general audience of the Julia community, not just the specific domain of users.

lu1and10 commented:
> @lu1and10 As the package author, would you be open to a longer and more descriptive name?
> 
> We usually try for package names that are descriptive to the general audience of the Julia community, not just the specific domain of users.

@DilumAluthge Thanks for asking. As one of the maintainers of the `FMM3D` package, personally I would prefer to keep the current name `FMM3D` agreeing with our friends @mipals @maltezfaria . We (the main authors https://fmm3d.readthedocs.io/en/latest/ackn.html) also discussed in our internal group meeting today, we do prefer to keep the current name if it is possible. We appreciate your help.
